Handle file rescan

WithoutPants opened the pull request:
- fixes issue where scene/image/gallery would not be created if it there was none associated with a file that had been subsequently moved
- scene/image/gallery post-update hooks are now called when the underlying file was updated
- scene generated files are now migrated when the underlying file hash has changed
- creates galleries when necessary instead of upfront when detecting a zip file. This fixes the issue where galleries would be created for zip files containing no valid image files
- clean now deletes empty galleries again
- fixes issue in clean where zip file folders would fail to be cleaned when its path is no longer valid because the zip file was moved
- fixes issue where duplicate images would not be associated with galleries - fixes #2939

djbtwothousandtwenty commented:
- first finding: the cleaning of empty galleries is not restricted to the "clean selected" path, it also cleaned a bunch of empty galleries outside of the path i selected


- second: it cleaned 3 empty galleries it made when encountering duplicate files previously, but on re-scan, it does not create new galleries for these duplicates. not sure if the folder paths need to be changed for them to be picked up, but right now it appears it's happy with the fact that the folders are in my folders table in the db, even if there's no gallery for the folder where there should be. i think if i change the folder name, it probably will pick it up as then the entry in the folders table "does not exist" for that specific path, so i'll try that for one of the dupes that's missing

it does pick that, and identifies that the images have been "moved" from old foldername to new foldername,  but again does not create the additional gallery. i'm now removing the performer folder alltogether, running clean to remove all these files/galleries/scenes, and then let it discover the whole performer folder again like it's never seen it before

this works, so when stash has never seen the content before, and does not have any path info in the folders or files tables, it created everything as i'd expect. No empty galleries, and 6 galleries with the correct names, the images are correctly connected to both the original as well as the "duplicate" gallery, and they show up correctly

the logic for the existing situation where the folders have previously been scanned and added to the folders table, but with an empty gallery created, leads to a weird state where it knows about the folder, it knows about the images, but it does not create the expected gallery

so i think the logic probably needs to check when it encounters a folder with images that would trigger the "create folder based gallery", whether it has one for that folder or not. if it does have one, update where required, if it does not have one, create a gallery and add the images found to it as it does now

- third finding might not be a bug at all: i know i've got a total of 5144 images in that specific performer folder. everything's tagged correctly, but when going to the images view, it reports a total of 4499 images. so that query "excludes duplicates". if i go filter for all images with a file count higher than 1, i get 639 duplicate images. which brings the total up to almost the correct number (4499+639=5138) which is still 6 short. this is because some cover files are duplicated in more than 2 folders, one of them is in 5 galleries for example. so i know by checking manually, all images are in the db, but maybe it would be good to check the files table for this performer for image filecount, and ignore whether duplicates are in the total? that way the imagecount would match the filesystem count

in this case, the reason for the duplicates is that it's very old met-art content, they were a bit messy in 2003, so some galleries got re-issued with different names, on different dates and sometimes with exactly the same photo's, sometimes with just a small difference in photocount, and some times with new cover files, sometimes with old ones. i could go and delete these dupes of course, but for me as a collector: this is how they released it, so that's how i would like to keep it, as close to the original as possible

WithoutPants commented:
Investigating this: I added two folders: `20220928` and `20220928.2`. I added three images to both folders, and an extra three to the latter folder. Scanning appears to be behaving correctly with regard to the image/gallery association. Duplicate files have a single image with both files associated, and associated to both folder galleries.

It is possible to add/remove images to/from folder/zip-based galleries, so I will plug that hole. Otherwise, unless I manually modify the database, it shouldn't be normally possible to get into the situation you describe, as far as I can tell. For scenarios where this has occurred, the recourse would probably be to either delete the folder-based gallery and rescan, or touch the affected image files, which should trigger a rescan and re-association.

I don't want to add a check to ensure that all images are associated to their correct galleries because it will add unnecessary overhead to scanning, it shouldn't be normally possible to get into this situation, and it is possible to workaround it if someone does get into the situation.
